unixfs/src/file/reader.rs

```diff
@@ -0,0 +1,333 @@
+    fn from_continued(
+        traversal: Traversal,
+        offset: u64,
+        data: &'a [u8],
+    ) -> Result<Self, FileReadFailed> {
+        let inner = FlatUnixFs::try_from(data)?;
+
+        if inner.data.mode.is_some() || inner.data.mtime.is_some() {
+            let metadata = FileMetadata::from(&inner.data);
+            return Err(FileError::NonRootDefinesMetadata(metadata))?;
+        }
+
+        Self::from_parts(inner, offset, traversal.metadata)
+    }
+
+    fn from_parts(
+        inner: FlatUnixFs<'a>,
+        offset: u64,
+        metadata: FileMetadata,
+    ) -> Result<Self, FileReadFailed> {
+        let empty_or_no_content = inner
+            .data
+            .Data
+            .as_ref()
+            .map(|cow| cow.as_ref().is_empty())
+            .unwrap_or(true);
+        let is_zero_bytes = inner.data.filesize.unwrap_or(0) == 0;
+
+        if inner.data.Type != UnixFsType::File && inner.data.Type != UnixFsType::Raw {
+            Err(FileReadFailed::UnexpectedType(inner.data.Type.into()))
+        } else if inner.links.len() != inner.data.blocksizes.len() {
+            Err(FileError::LinksAndBlocksizesMismatch)?
+        } else if empty_or_no_content && !is_zero_bytes && inner.links.is_empty() {
+            Err(FileError::NoLinksNoContent)?
+        } else {
+            // raw and file seem to be same except the raw is preferred in trickle dag
+            let data = inner.data.Data.unwrap_borrowed_or_empty();
+
+            if inner.data.hashType.is_some() || inner.data.fanout.is_some() {
+                return Err(FileError::UnexpectedRawOrFileProperties {
+                    hash_type: inner.data.hashType,
+                    fanout: inner.data.fanout,
+                }
+                .into());
+            }
+
+            let end = if inner.links.is_empty() {
+                // can unwrap because `data` is all of the data
+                let filesize = inner.data.filesize.unwrap_or(data.len() as u64);
+                Ending::Chunk(offset + filesize)
+            } else {
+                match inner.data.filesize {
+                    Some(filesize) => Ending::TreeCoverage(offset + filesize),
+                    None => return Err(FileError::IntermediateNodeWithoutFileSize.into()),
+                }
+            };
+
+            Ok(Self {
+                offset,
+                end,
+                had_links: !inner.links.is_empty(),
+                links: inner.links,
+                data,
+                blocksizes: inner.data.blocksizes,
+                metadata,
+            })
+        }
+    }
+
+    /// Returns a moved tuple of the content (bytes or links) and a traversal, which can be used to
+    /// continue the traversal from the next block.
+    pub fn content(
+        self,
+    ) -> (
+        FileContent<'a, impl Iterator<Item = (PBLink<'a>, Range<u64>)>>,
+        Traversal,
+    ) {
+        let traversal = Traversal {
+            last_had_links: !self.had_links,
```

Comment:
This looks fishy.

Reply:
Double negative but no test failures is triggered by changing only one of these. Turns out these were never needed but they were used enough not to trigger unused field warnings. Removing.